feat(track): add `composer` and `composer_sort` fields

ali-ramadhan opened the pull request:
- [x] I have read the contributing guide in the documentation.

### Description

This update introduces two new fields, `composer` and `composer_sort`, to tracks. 

It's my attempt at addressing the issue raised by discussion #282 around adding more track and album metadata. If we can add `composer` and `composer_sort` then we should be able to tag classical music better and I'll have a better understanding of how to add more metadata.

I assume I should also include/generate an Alembic migration, something like

```python
def upgrade():
    with op.batch_alter_table("track", schema=None) as batch_op:
        batch_op.add_column(sa.Column("composer", sa.String(), nullable=True))
        batch_op.add_column(sa.Column("composer_sort", sa.String(), nullable=True))


def downgrade():
    with op.batch_alter_table("track", schema=None) as batch_op:
        batch_op.drop_column("composer_sort")
        batch_op.drop_column("composer")
```
?


<!-- readthedocs-preview mrmoe start -->
----
📚 Documentation preview 📚: https://mrmoe--295.org.readthedocs.build/en/295/

<!-- readthedocs-preview mrmoe end -->

ali-ramadhan commented:
Ah haha I think I massacred the tests by not including the migration yet.

jtpavlock commented:
Not sure why I didn't realize this before, but there's already a basic alembic section in the [contributing docs](https://mrmoe.readthedocs.io/en/latest/developers/contributing.html#migrations) that should help you get started with this. Let me know if you have any other questions.

ali-ramadhan commented:
Thanks for linking to the docs (which I should have been familiar with lol)! Was able to create the alembic migration and the new field checklist was very useful!

Should we be adding the `composer_sort` field in this PR? Although it may be a common tag, `artist_sort` is actually not a `Track` field. So maybe it makes more sense for this PR to only add the `composer` field? If we want `artist_sort` and `composer_sort` they can be introduced together in a future PR? Alternatively, we can merge this with `composer_sort` and I'm happy to open a PR to also add `artist_sort`.

Hmmm, so after adding `composers` pyright told me that `MediaFile` actually does not support `composers`. Confusingly (to me, at least) there are fields for `artist`, `artists`, and `artist_sort`, but only `composer` and `composer_sort`: https://github.com/beetbox/mediafile/blob/master/mediafile.py

How do you think we should proceed here?

ali-ramadhan commented:
Thanks for switching this to a draft PR, makes sense for now while we discuss!

I'm thinking it might be best to focus this PR on just adding the `composer` and `composer_sort` field since they are both supported by mediafile so such a PR would be straightforward while also adding some useful fields that I would be interested in using immediately.

If I go ahead with just `composer` and `composer_sort` would you prefer one commit with both, or one commit for `composer` and another for `composer_sort`?

I don't mind opening a mediafile PR adding a `composers` field, but from reading mediafile issues it seems like the "composers" tag would have to be supported by at least some music players to warrant addition. And I'm not sure I could find any info on this.

jtpavlock commented:
I'd prefer each new field to be its own `feat` commit with its own migration script and changes. Makes sense about the mediafile change though. Like I said we could always maintain the multiple values in the database, but if you don't think that'd be useful just doing the singular version 'composer' is fine and we can always migrate to a multi-value field later on.

ali-ramadhan commented:
Alright I think this is ready for review! This PR now just adds the `composer` and `composer_sort` fields to `Track`. There is one commit for `composer` first then another for `composer_sort`.

jtpavlock reviewed (changes requested):
This is wonderful, very nice job here. The only thing I think we should add is a weight for composers to be matched in match.py. I have no immediately strong opinions for how much they should be weighted as I don't really use composers in my library so over to you what you think would be good.

I also added a few commits to update the new field checklist and a bit of the related tests. I made one minor change to your code too, I just removed the merge and overwrite tests only because our implementation solution doesn't hardcode any fields in so there's no code change for a new field, and I don't currently test to make sure every field merges properly so I'm just keeping it consistent for now. I apologize also for the force push to your branch.

ali-ramadhan commented:
Thank you for the kind words @jtpavlock! I'm very happy to join as a maintainer. I am interested in continuing to use Moe and I hope it becomes a popular and mature package! And I do enjoy discussing/reviewing code and answering questions. I also think you maintain an excellent software package and I'm actively learning how to code and use git better here. I may not always have tons of time but I am on GitHub almost every day as part of my job so I can try to contribute in a timely manner when I'm not super busy. And if I feel like I can't contribute anymore I will let you know.

Thanks for the changes! I just added a fixup commit to add a weight for composers. Since it's just a `Track` field I added it to `MATCH_TRACK_FIELD_WEIGHTS` with a weight of 0.8 (same as album artists). Seems like a reasonable default to me.

jtpavlock commented:
Thank you for your kind words in return! I do think I can be overly picky about code sometimes, but I'm glad you're taking it as constructive. I'll also go ahead and invite you to the organization, thanks for being willing to take on a more involved role. Once you accept, I should be able to give you write permissions. 

Also, let me know if you think it would be worthwhile to have a separate chat for more informal or quick discussions outside GitHub. I've used [zulip](https://zulip.com/) in the past as an open-source-friendly option that seems to work well, but I'm also open to alternatives.

ali-ramadhan commented:
Thanks for the invite @jtpavlock, I'm in!

Having a space to quickly chat can be helpful! It certainly enables more conversations as it lowers the barrier for communication (writing up a new GitHub issue or discussion is a lot more work than just posting a chat message). I haven't used Zulip in a few years but I remember it was pretty good. Out of curiosity, do you also see this space as being for users with questions as well?

jtpavlock commented:
Awesome! You should have write privileges now. I'll also go ahead and invite you to a chat I just created on zulip, and we can later decide if we'd rather use something else. I do think the chat should only be for maintainers though, I think user issues/questions should be handled via github discussions and issues.

ali-ramadhan commented:
> I do think the chat should only be for maintainers though, I think user issues/questions should be handled via github discussions and issues.

I think a strong argument for handling user issues/questions via GitHub only is that the discussions and solutions become easily searchable and increase visibility for the package online. On Slack they can just get lost in time.
